helix-core/src/main/java/org/apache/helix/controller/stages/TopStateHandoffReportStage.java

```diff
@@ -494,6 +494,9 @@ private void reportTopStateComesBack(ResourceControllerDataProvider cache, Map<S
 
     // In case of recovery after failure, we should decrement the missingTopStateBeyondThresholdGauge value.
     if (clusterStatusMonitor != null && record.isFailed()) {
+      LogUtil.logInfo(LOG, _eventId, String.format(
+          "Missing top state recovered for resource %s and partition %s. Decrementing missingTopStateBeyondThresholdGauge.",
+          resourceName, partition.getPartitionName()));
```

Comment:
Top state comes back is pretty frequent.  Will we flood log?

Reply:
This wont be published all the time, and will be published only when the threshold is missed (and the record is set to failed).  
We are publishing similar log during increment as well, to ensure that metric is emitted, given this is a very critical metric for end-users.